heartbeat/README.md

```diff
@@ -0,0 +1,176 @@
+can't make progress with 1/3 of nodes offline at any given time for a nightly rest.
+
+To combat this scenario, the network generates random _challenges_ periodically.  Every
+`Payouts.ChallengeInterval` rounds (currently 1000), a random selected portion (currently 1/32) of
+all online accounts are challenged.  They must _heartbeat_ within `Payouts.ChallengeGracePeriod`
+rounds (currently 200), or they will be subject to suspension. With the current consensus
+parameters, nodes can be expected to be challenged daily.  When suspended, accounts must `keyreg`
+with the `GoOnlineFee` in order to receive block payouts again, so it becomes unprofitable for
+these low-stake nodes to operate with poor uptimes.
+
+# Heartbeats
+
+The absenteeism mechanism is subject to rare false positives.  The challenge mechanism explicitly
+requires an affirmative response from nodes to indicate they are operating properly on behalf of a
+challenged account.  Both of these needs are addressed by a new transaction type --- _Heartbeat_. A
+Heartbeat transaction contains a signature (`HbProof`) of a recent block seed (`HbSeed`) under the
+participation key of the account (`HbAddress`) in question. Note that the account being heartbeat
+for is _not_ the `Sender` of the transaction, which can be any address.
+
+It is relatively easy for a bad actor to emit Heartbeats for its accounts without actually
+participating. However, there is no financial incentive to do so.  Pretending to be operational when
+offline does not earn block payouts.  Furthermore, running a server to monitor the block chain to
+notice challenges and gather the recent blockseed is not significantly cheap that simply running a
+functional node. It is _already_ possible for malicious, well-resourced accounts to cause consensus
+difficulties by going online without actually participating.  Heartbeats do not mitigate that
+risk. But these mechanisms have been designed to avoid _motivating_ such behavior, so that they can
+accomplish their actual goal of noticing poor behavior stemming from _inadvertent_ operational
+problems.
+
+## Free Heartbeats
+
+Challenges occur frequently, so it important that `algod` can easily send Heartbeats as
+required. How should these transactions be paid for? Many accounts, especially high-value accounts,
+would not want to keep their spending keys available for automatic use by `algod`. Further, creating
+(and keeping funded) a low-value side account to pay for Heartbeats would be an annoying operational
+overhead.  Therefore, when required by challenges, heartbeat transactions do not require a fee.
+Therefore, any account, even an unfunded logigsig, can send heartbeats for an account under
+challenge.
+
+The conditions for a free Heartbeat are:
+
+1. The Heartbeat is not part of a larger group, and has a zero `GroupID`.
+1. The `HbAddress` is Online and under challenge with the grace period at least half over.
```

Comment:
I figured it would be nice to let things play out for a little while first. Any challenged node that proposes during the first half of the grace period doesn't have to heartbeat.  I suppose that's not likely to be common. 1/32 accounts are challenged - so maybe we get three nodes heartbeating during the 100 rounds that constitute half a grace period?

We could increase the grace period (at most 500, it must be less than half our ability to lookup headers, which is 1001).  That would delay suspensions, but who cares?  We could also make it 3/4 of grace period?  Or grace period minus some constant that gives nodes enough time to heartbeat. 50?

Or just leave it. I don't have any strong feelings.  But I will add a unit test to confirm nobody ever makes the grace period more than half of our header lookup ability.